include/aws/s3/private/s3_client_impl.h

```diff
@@ -21,113 +22,104 @@
 
 struct aws_http_connection;
 struct aws_http_connection_manager;
-
-typedef void(aws_s3_client_acquire_http_connection_callback)(
-    struct aws_http_connection *http_connection,
-    int error_code,
-    void *user_data);
-
-typedef void(aws_s3_client_sign_callback)(int error_code, void *user_data);
+struct aws_host_resolver;
+struct aws_s3_endpoint;
 
 enum aws_s3_vip_connection_finish_code {
     AWS_S3_VIP_CONNECTION_FINISH_CODE_SUCCESS,
     AWS_S3_VIP_CONNECTION_FINISH_CODE_FAILED,
     AWS_S3_VIP_CONNECTION_FINISH_CODE_RETRY,
 };
 
+typedef void(aws_s3_client_sign_callback)(int error_code, void *user_data);
+
 typedef void(aws_s3_vip_shutdown_callback_fn)(void *user_data);
 
-/* Represents one Virtual IP (VIP) in S3, including a connection manager that points directly at that VIP. */
-struct aws_s3_vip {
-    struct aws_linked_list_node node;
+typedef bool(aws_s3_endpoint_ref_zero_fn)(struct aws_s3_endpoint *endpoint);
 
-    /* True if this VIP is in use. */
-    struct aws_atomic_var active;
+typedef void(aws_s3_endpoint_shutdown_fn)(void *user_data);
 
-    /* S3 Client that owns this vip. */
-    struct aws_s3_client *owning_client;
+struct aws_s3_endpoint_options {
+    /* URL of the host that this endpoint refers to. */
+    struct aws_string *host_name;
 
-    /* Connection manager shared by all VIP connections. */
-    struct aws_http_connection_manager *http_connection_manager;
+    /* Callback for when the reference count of this endpoint hits zero. */
+    aws_s3_endpoint_ref_zero_fn *ref_count_zero_callback;
```

Comment:
just call this a destructor, I know it's nomenclature that aws_refcount uses, but this is weird to read